nx/source/services/hid.c

```diff
@@ -69,6 +106,135 @@ void* hidGetSharedmemAddr(void) {
     return shmemGetAddr(&g_hidSharedmem);
 }
 
+void hidSetControllerLayout(HIDControllerID id, HIDControllerLayoutType layoutType) {
+    controllerLayout[id] = layoutType;
+}
+
+void hidScanInput(void) {
+    if (g_hidServiceSession == INVALID_HANDLE) return;
+    HIDSharedMemory *sharedMem = (HIDSharedMemory*)hidGetSharedmemAddr();
+
+    mouseOld = mouseHeld;
+    keyboardModOld = keyboardModHeld;
+    memcpy(keyboardOld, keyboardHeld, sizeof(keyboardOld));
+    memcpy(controllerOld, controllerHeld, sizeof(controllerOld));
+
+    mouseHeld = 0;
+    keyboardModHeld = 0;
+    memset(keyboardHeld, 0, sizeof(keyboardHeld));
+    memset(controllerHeld, 0, sizeof(controllerHeld));
+    memset(&touchEntry, 0, sizeof(HIDTouchScreenEntry));
+    memset(&mouseEntry, 0, sizeof(HIDMouseEntry));
+    memset(&keyboardEntry, 0, sizeof(HIDKeyboardEntry));
+    for (int i = 0; i < 10; i++)
+        memset(&controllerEntries[i], 0, sizeof(HIDControllerInputEntry));
+
+    u64 latestTouchEntry = sharedMem->touchscreen.header.latestEntry;
+    HIDTouchScreenEntry *newTouchEntry = &sharedMem->touchscreen.entries[latestTouchEntry];
+    if (newTouchEntry->header.timestamp - touchTimestamp) {
```

Comment:
What's going on with this if statement

Reply:
I think it should be rewritten to `if ((s64)(newTouchEntry->header.timestamp - touchTimestamp) >= 0)`. It is meant to be a basic greater-than check, accounting for the fact that the timestamp overflowing the 64-bit variable is technically possible and valid.

Reply:
Right now the comparison is equivalent to (newTouchEntry->header.timestamp != touchTimestamp) afaict

Reply:
It is. But of course, incoming touch entries with a previous timestamp would also pass the check... Not sure if we want to assume that will never happen...

Reply:
Rollover is unlikely, but I think I'll go ahead and move to a greater-than zero compare for clarity.